File: backend/app/models/power.py
from datetime import datetime
from pydantic import BaseModel
from typing import Optional


class PowerReading(BaseModel):
    """Single power reading at a point in time"""
    timestamp: float
    watts: float
    voltage: Optional[float] = None
    current: Optional[float] = None
    source: str  # 'battery', 'ac_adapter', 'ipmi', 'estimated'


class PowerMetrics(BaseModel):
    """Current power metrics for WebSocket/API"""
    timestamp: float
    watts: float
    voltage: Optional[float] = None
    current: Optional[float] = None
    source: str
    total_kwh: float  # Accumulated since logging started
    uptime_hours: float


class PowerHistoryRequest(BaseModel):
    start_date: datetime
    end_date: datetime


class PowerHistoryResponse(BaseModel):
    start_date: datetime
    end_date: datetime
    total_kwh: float
    avg_watts: float
    max_watts: float
    min_watts: float
    readings_count: int
    hours_span: float


class PowerReadingEntry(BaseModel):
    timestamp: float
    watts: float

File: frontend/src/types/power.ts
export interface PowerMetrics {
  timestamp: number;
  watts: number;
  voltage: number | null;
  current: number | null;
  source: string;
  total_kwh: number;
  uptime_hours: number;
}

export interface PowerHistoryResponse {
  start_date: string;
  end_date: string;
  total_kwh: number;
  avg_watts: number;
  max_watts: number;
  min_watts: number;
  readings_count: number;
  hours_span: number;
}

export interface PowerReadingEntry {
  timestamp: number;
  watts: number;
}

export interface RecentPowerResponse {
  readings: PowerReadingEntry[];
  count: number;
}
